pkg/agent/plugin/keymanager/base/keymanagerbase.go

```diff
@@ -0,0 +1,338 @@
+package keymanagerbase
+
+import (
+	"context"
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/sha256"
+	"crypto/x509"
+	"encoding/hex"
+	"fmt"
+	"sort"
+	"sync"
+
+	keymanagerv1 "github.com/spiffe/spire-plugin-sdk/proto/spire/plugin/agent/keymanager/v1"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/proto"
+)
+
+// KeyEntry is an entry maintained by the key manager
+type KeyEntry struct {
+	PrivateKey crypto.Signer
+	*keymanagerv1.PublicKey
+}
+
+// Funcs is a collection of optional callbacks. Default implementations will be
+// used when not provided.
+type Funcs struct {
+	WriteEntries       func(ctx context.Context, allEntries []*KeyEntry, newEntry *KeyEntry) error
+	GenerateRSA2048Key func() (*rsa.PrivateKey, error)
+	GenerateRSA4096Key func() (*rsa.PrivateKey, error)
+	GenerateEC256Key   func() (*ecdsa.PrivateKey, error)
+	GenerateEC384Key   func() (*ecdsa.PrivateKey, error)
+}
+
+// Base is the base KeyManager implementation
+type Base struct {
+	keymanagerv1.UnsafeKeyManagerServer
+	funcs Funcs
+
+	mu      sync.RWMutex
+	entries map[string]*KeyEntry
+}
+
+// New creates a new base key manager using the provided Funcs. Default
+// implementations are provided for any that aren't set.
+func New(funcs Funcs) *Base {
+	if funcs.GenerateRSA2048Key == nil {
+		funcs.GenerateRSA2048Key = generateRSA2048Key
+	}
+	if funcs.GenerateRSA4096Key == nil {
+		funcs.GenerateRSA4096Key = generateRSA4096Key
+	}
+	if funcs.GenerateEC256Key == nil {
+		funcs.GenerateEC256Key = generateEC256Key
+	}
+	if funcs.GenerateEC384Key == nil {
+		funcs.GenerateEC384Key = generateEC384Key
+	}
+	return &Base{
+		funcs:   funcs,
+		entries: make(map[string]*KeyEntry),
+	}
+}
+
+// SetEntries is used to replace the set of managed entries. This is generally
+// called by implementations when they are first loaded to set the initial set
+// of entries.
+func (m *Base) SetEntries(entries []*KeyEntry) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	m.entries = entriesMapFromSlice(entries)
+	// populate the fingerprints
+	for _, entry := range m.entries {
+		entry.PublicKey.Fingerprint = makeFingerprint(entry.PublicKey.PkixData)
+	}
+}
+
+// GenerateKey implements the KeyManager RPC of the same name.
+func (m *Base) GenerateKey(ctx context.Context, req *keymanagerv1.GenerateKeyRequest) (*keymanagerv1.GenerateKeyResponse, error) {
+	if req.KeyId == "" {
+		return nil, status.Error(codes.InvalidArgument, "key id is required")
+	}
+	if req.KeyType == keymanagerv1.KeyType_UNSPECIFIED_KEY_TYPE {
+		return nil, status.Error(codes.InvalidArgument, "key type is required")
+	}
+
+	newEntry, err := m.generateKeyEntry(req.KeyId, req.KeyType)
+	if err != nil {
+		return nil, err
```

Comment:
I noticed that there are several places where a similar thing happens, including the server side. I'm sorry I missed this in previous reviews.
I think that anything we can do to have errors that help to identify and fix issues is a good thing. I'm curious to know your thoughts.

Reply:
yes, for sure. I've added the ones you've pointed out and added some code to wrap the error generically with a prefix like "failed to generate key"